log-levels.md

```diff
@@ -0,0 +1,46 @@
+
+**// TODO: should we suggest that a system should be able to survive with debug logging turned on in production (ktoso: I'd suggest that, yes).**
+
+This is in contrast with `debug` which is some users _may_ choose to run with enabled on their production systems. As such, debug logging should not be overly noisy, and provide a high value understanding of what is going on in the library for end users, without diving deep into its internals (which `trace` is intended for).
+
+### Log levels to avoid
+
+All these rules are only _general_ guidelines, and as such may have exceptions to the rules. Please consider the following examples and rationale for why logging at high log levels by a library may not be desirable:
+
+It is generally _not acceptable_ for a service client (e.g. an http client) to log an `error` when a request has failed. End-users may be using the client to probe if an endpoint is even responsive or not, and a failure to respond may be _expected_ behavior for the end-user, and logging errors would only confuse and pollute their logs.
+Instead, libraries should either `throw`, or return an `Error` value that users of the library will have enough knowladge about if they should log or ignore it.
+
+It is even less acceptable for a library to log any successful operations. This leads to flodding server side systems, especially 
+
+### Suggested logging style
+
+Libraries may want to embrace the structured logging style.
+
+**TODO: fill this in more.**
+
+### Exceptions to the rules
+
+As this is only a general guideline, there are exceptions to this rule which must be decided on case-by-case depending on the library and situation at hand.
+
+Here are a few examples of situations when logging a message on a relatively high level might still be tolerable for a library.
+
+It permittable for a library to log at `critical` level right before a _hard crash_, as a last resort of informing the log collection systems or end-user about additional information detailing the reason for the crash, which _in addition to_ the message from a `fatalError` can lead to an improved diagnosis/debugging experience for end users.
+
+Sometimes libraries may be able to detect a harmful misconfiguration of the library. E.g. selecting deprecated protocol versions or similar. In such situations it may be useful to inform users on production by issuing a `warning`, however do take care that such warning is not issued repeatedly! For example, it is not acceptable for an HTTP client to log a warning on every single http request using some misconfiguration of the client. It _may_ be acceptable however for the client to log such warning e.g. _once_ at configuration time, if the library has a good way to do this.
```

Comment:
```suggestion
Sometimes libraries may be able to detect a harmful misconfiguration of the library. For example, selecting deprecated protocol versions. In such situations it may be useful to inform users in production by issuing a `warning`. However you should ensure that the warning is not logged repeatedly! For example, it is not acceptable for an HTTP client to log a warning on every single http request using some misconfiguration of the client. It _may_ be acceptable however for the client to log such warning e.g. _once_ at configuration time, if the library has a good way to do this.
```